phabricatoremails/init.py

```diff
@@ -0,0 +1,33 @@
+import pathlib
+
+from alembic import command
+from alembic.config import Config as AlembicConfig
+from phabricatoremails.settings import Settings
+
+
+def init(settings: Settings):
```

Comment:
Hmm, I don't follow the benefit of splitting this into three functions for a couple reasons:

1. The different steps you've laid out aren't separate-able in the code:
```
# Performed before we create the tables so that we fail early if Phabricator can't be reached
end_key = source.fetch_feed_end()  # step (3)
# Performed before we set_initial_position() because we need tables before we can insert rows
command.upgrade(AlembicConfig(...), "head")  # step (1)
progress.set_initial_position(db_session, end_key)  # step (3)
```

2. What benefit will we receive by separating these? This high-level code is already only testable in an "end-to-end" test fashion (the actual logic of fetching `end_key`, doing `set_initial_position()`, etc are done at lower levels). If they're not being separated to reduce scope for testing, then what's beneficial about:

```
def create_db(alembic_config):
    command.upgrade(alembic_config, "head")

#  (I'm ignoring the staggering of logic for this example)
def fetch_and_set_position(db_session, source, progress):
    end_key = source.fetch_feed_end()
    progress.set_initial_position(db_session, end_key)
    return end_key

def init(settings):
    db = settings.db()
    source = settings.source
    progress = settings.progress

    if db.is_initialized():
        # log and return
    create_db(AlembicConfig(...))
    end_key = fetch_and_set_position(db_session, source, progress)

    # log end_position
```

I understand that there's general guidelines to "keep functions short" and "functions should only have a single identifiable purpose", but I balanced that against over-verbosity and the constraints. If you're recommending this refactor to improve maintainability, can you clarify how this would help?

Reply:
> The different steps you've laid out aren't separate-able in the code:

The migration step is _repeated_ elsewhere in the code, in a context that makes sense (i.e. `migrate.py`). It is generally best to not repeat code whenever possible.

I would also raise a noisy exception if someone tries to run this method when the database is already initialized, rather than just skipping.

https://github.com/mozilla-conduit/phabricator-emails/pull/1/files#r407523722

Reply:
Good call on the exception, I'll set that up.

> The migration step is repeated elsewhere in the code, in a context that makes sense (i.e. migrate.py). It is generally best to not repeat code whenever possible.

Ah, I see. For this specific case, I've added a response [over here](https://github.com/mozilla-conduit/phabricator-emails/pull/1/files#r407691781).

> And to clarify I did suggest separating this into 2 methods, not 3 (i.e. the "initializing" and the "migrating" though you already have the migrating part done elsewhere -- sorry if this was confusing, but I assume that this "init" method only runs once per fresh local install, and only once in a production environment ever)

I'm a little confused here, but I thiiiink we're on the same page? You're right that this `init()` method should run once per install (then afterwords `migrate()` is used each time the code changes). I'm not sure which other method you had in mind to extract, and perhaps the discussion linked above^ will still apply?

Reply:
What I'm suggesting here is the below, since lines 21-22 are repeated in `phabricatoremails.migrate`. There's also some good logging there that is relevant.

```python
from phabricatoremails.migrate import migrate. # at the top of the file
migrate(settings)
```

Reply:
I don't think that `init()` should call `migrate()`, since the work we want `init()` to do isn't a superset of what we want `migrate()` to do.
For example, `migrate()` verifies that the database is initialized (this is unnecessary when called from `init()`), and it prints logs according to the "migration" use case.

In the future, the overlap between `init()` and `migrate()` will probably decrease further - for example, we could send `statsd` information upon a migration happening.

This "superset" relationship between `init()` and `migrate()` seems superficial/surface-level to me. Having `init()` call `migrate()` will reduce the raw number of lines of code, but will increase friction to `migrate()` evolving independently.

Reply:
I'm not sure "superset" is an accurate descriptor here. I think of these more as "sequential". I.e., "initializing" will do perform "migrate" and then do some other stuff. Migrate shouldn't care whether we've "initialized" the database or not, it should just look at the current version of the database, and migrate if necessary. `upgrade` here in this module simply doesn't do anything if you've ran `migrate`, and `migrate` doesn't cause any problems if it runs without this `init` having run first.

Reply:
> Migrate shouldn't care whether we've "initialized" the database or not, it should just look at the current version of the database, and migrate if necessary. 

But, calling `phabricator-emails migrate` _shouldn't_ create and migrate the database if it hasn't been created yet. This is because initializing the database requires setting the current position in the Phabricator feed, and that should only happen once (on initialization).

-----

`init()` and `migrate()` do share some logic (getting the alembic config, performing the alembic db upgrade), but they're otherwise distinct, and will continue to evolve separately.

Reply:
FWIW, in the [this commit](https://github.com/mozilla-conduit/phabricator-emails/pull/1/commits/5447266e90630f5040347473b371c064844dd729), I've moved the schema-upgrade logic into DB.